File: packages/coding-agent/src/core/extensions/subagent-reserved-names.ts
import type { ExtensionFactory } from "./types.js";

const SUB_AGENT_RESERVED_PREFIX = "sub_agent.";
const SUB_AGENT_RESERVED_NAMES = new Set([
	"sub_agent.delegate",
	"sub_agent.readiness",
	"sub_agent.delegation.result",
	"sub_agent.status",
	"sub_agent_readiness",
	"sub-agent",
	"/sub-agent",
]);

const SUB_AGENT_EXTENSION_FACTORY_BRAND: unique symbol = Symbol.for("pi.subAgentExtensionFactory");

type SubAgentExtensionFactory = ExtensionFactory & {
	[SUB_AGENT_EXTENSION_FACTORY_BRAND]?: true;
};

export function markSubAgentExtensionFactory(factory: ExtensionFactory): ExtensionFactory {
	Object.defineProperty(factory, SUB_AGENT_EXTENSION_FACTORY_BRAND, {
		value: true,
		configurable: false,
		enumerable: false,
		writable: false,
	});
	return factory;
}

export function isSubAgentExtensionFactory(factory: ExtensionFactory): boolean {
	return (factory as SubAgentExtensionFactory)[SUB_AGENT_EXTENSION_FACTORY_BRAND] === true;
}

export function isSubAgentReservedName(name: string): boolean {
	return SUB_AGENT_RESERVED_NAMES.has(name) || name.startsWith(SUB_AGENT_RESERVED_PREFIX);
}

export function assertSubAgentReservedNameAllowed(
	name: string,
	ownerAllowed: boolean | undefined,
	operation: string,
): void {
	if (!isSubAgentReservedName(name) || ownerAllowed === true) {
		return;
	}
	throw new Error(`Cannot ${operation} reserved sub-agent substrate name "${name}" from an unrelated extension.`);
}

File: zig/src/coding_agent/core/extensions/subagent_reserved_names.zig
const std = @import("std");

pub const SUB_AGENT_RESERVED_PREFIX = "sub_agent.";
pub const SUB_AGENT_RESERVED_NAMES = [_][]const u8{
    "sub_agent.readiness",
    "sub_agent.delegation.result",
    "sub_agent.status",
    "sub_agent.delegate",
};

pub const SubAgentReservedNameError = error{SubAgentReservedNameDenied};

pub const SubAgentExtensionFactory = struct {
    is_sub_agent_factory: bool = false,
};

pub fn markSubAgentExtensionFactory(factory: *SubAgentExtensionFactory) *SubAgentExtensionFactory {
    factory.is_sub_agent_factory = true;
    return factory;
}

pub fn isSubAgentExtensionFactory(factory: *const SubAgentExtensionFactory) bool {
    return factory.is_sub_agent_factory;
}

pub fn isSubAgentReservedName(name: []const u8) bool {
    if (std.mem.startsWith(u8, name, SUB_AGENT_RESERVED_PREFIX)) return true;
    inline for (SUB_AGENT_RESERVED_NAMES) |reserved| {
        if (std.mem.eql(u8, name, reserved)) return true;
    }
    return false;
}

pub fn assertSubAgentReservedNameAllowed(
    name: []const u8,
    owner_is_sub_agent_factory: bool,
    operation: []const u8,
) SubAgentReservedNameError!void {
    _ = operation;
    if (!owner_is_sub_agent_factory and isSubAgentReservedName(name)) return error.SubAgentReservedNameDenied;
}

test "sub-agent reserved names use prefix contract" {
    try std.testing.expect(isSubAgentReservedName("sub_agent.delegate"));
    try std.testing.expect(isSubAgentReservedName("sub_agent.custom"));
    try std.testing.expect(!isSubAgentReservedName("agent.delegate"));
}

test "sub-agent factory marker authorizes reserved names" {
    var factory: SubAgentExtensionFactory = .{};
    try std.testing.expect(!isSubAgentExtensionFactory(&factory));
    _ = markSubAgentExtensionFactory(&factory);
    try std.testing.expect(isSubAgentExtensionFactory(&factory));

    try std.testing.expectError(
        error.SubAgentReservedNameDenied,
        assertSubAgentReservedNameAllowed("sub_agent.delegate", false, "register"),
    );
    try assertSubAgentReservedNameAllowed("sub_agent.delegate", true, "register");
}
